fix(core): `QueryList` does not fire changes if the underlying list did not change.

mhevery opened the pull request:
Previous implementation would fire changes `QueryList.changes.subscribe`
even if underlying list did not change. Such situation can arise if a
`LView` is inserted or removed, causing `QueryList` to recompute, but
the recompute results in the same exact `QueryList` result. In such
a case we don't want to fire a change event (as there is no change.)

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery committed:
fix(core): `QueryList` should not fire changes if the underlying list did not change.

Previous implementation would fire changes `QueryList.changes.subscribe`
whenever the `QueryList` was recomputed. This resulted in artificially
high number of change notifications, as it is possible that recomputing
`QueryList` results in the same list. When the `QueryList` gets recomputed
is an implementation detail and it should not be the thing which determines
how often change event should fire.

This change introduces a new `emitDistinctChangesOnly` option for
`ContentChildren` and `ViewChildren`.

```
export class QueryCompWithStrictChangeEmitParent {
  @ContentChildren('foo', {
    // This option will become the default in the future
    emitDistinctChangesOnly: true,
  })
  foos!: QueryList<any>;
}
```

mhevery commented:
[presubmit](http://test/OCL:347030270:BASE:347030299:1607712420574:af459fcf)

mhevery commented:
[presubmit](http://test/OCL:347030270:BASE:347055259:1607719470143:d65b411d)

IgorMinar reviewed (changes requested):
Overall this looks good, thanks for debugging this.

Could you instead consider using the native [Array#flat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) as this would simplify the code and in fact remove bulk of it.

We'll then also need to add

```
import 'core-js/modules/es.array.flat';
import 'core-js/modules/es.array.flat-map';
```
to

https://github.com/angular/angular-cli/blob/0d10de5cbba9f4eaf91e0fbc279a1f3e0ad94fa3/packages/angular_devkit/build_angular/src/webpack/es5-polyfills.js#L36-L41

just so that we don't break on IE11 (the flat-map addition is not required by this change, but we might as well add it for future use).

If we target this fix for `target: minor` then we can roll it out now since CLI would add the polyfill to ie11 builds in v11.1 as well.

AndrewKushnir commented on packages/core/src/util/array_utils.ts:
```diff
@@ -24,22 +24,62 @@ export function addAllToArray(items: any[], arr: any[]) {
+ *
+ * @param src Source array which can be hierarchical (array within array)
+ * @param dst Destination array which can have existing values.
+ * @returns `true` if `dst` array was muted in order to represent flatten version of `src`.
```
```suggestion
 * @returns `true` if `dst` array was mutated in order to represent flatten version of `src`.
```

mhevery replied:
fixed

AndrewKushnir commented on packages/core/test/acceptance/query_spec.ts:
```diff
@@ -1123,6 +1138,42 @@ describe('query logic', () => {
+      });
+
+      // First verify that the subscription is working.
+      fixture.componentInstance.innerShowing = false;
```
I was a bit confused by the fact that we always have 1 item in `foos` and I realized that I was missing this flag :)
I'd propose to add a couple comments to explain this.

mhevery replied:
Can you suggest the exact comment you would like to see.

AndrewKushnir replied:
My comment seems obsolete now since you've updated tests and added a comment at the very end of the test.

AndrewKushnir commented on packages/core/test/acceptance/query_spec.ts:
```diff
@@ -1123,6 +1138,42 @@ describe('query logic', () => {
+      // now verify that adding a view does not needlessly fire subscription
+      fixture.componentInstance.showing = true;
+      fixture.detectChanges();
+      expect(changes).toBe(isVEQueryBug ? 3 : 1);
+      expect(componentInstance.foos.toArray().length).toEqual(1);
```
```suggestion
      expect(changes).toBe(isVEQueryBug ? 3 : 1);
      // Note: even though the `showing` is `true` and the second `<div>` is displayed, the child element 
      // of that <div> is hidden because the `innerShowing` flag is still `false`, so we expect only one element
      // to be present in the `foos` array.
      expect(componentInstance.foos.toArray().length).toEqual(1);
```

mhevery replied:
done

AndrewKushnir reviewed (approved):
LGTM (from code perspective), thanks for the fix @mhevery 👍 

I was also thinking of using `Array.flat` function as @IgorMinar proposed, but that'd result in additional array allocation and extra pass to compare current/new arrays (which is avoided in the current implementation). Given that `QueryList.reset` is called for each CD run (executed from `refreshView` fn), it looks like that might sensitive from perf perspective...

JoostK commented on packages/core/src/util/array_utils.ts:
```diff
@@ -24,22 +24,62 @@ export function addAllToArray(items: any[], arr: any[]) {
+  let dstIndx = flattenIntoExistingRecursive(src, dst, 0);
+  let differenceDetected = dstIndx < 0;
+  dstIndx = differenceDetected ? ~dstIndx : dstIndx;
+  while (dstIndx < dst.length) {
+    dst.pop();
+    differenceDetected = true;
+  }
```
Could the following be faster?

```suggestion
  if (dstIndx < dst.length) {
    // Fewer items than were previously in `dst`, so trim the array.
    dst.length = dstIndex;
    differenceDetected = true;
  }
```

mhevery replied:
fixed

mhevery commented on packages/core/src/linker/query_list.ts:
```diff
@@ -135,13 +135,20 @@ export class QueryList<T> implements Iterable<T> {
    *
    * @param resultsTree The query results to store
+   * @returns `true` only if the new `resultTree` resulted in a different query set.
+   *          (It could be that a new `resultTree` produces the same sat in which case we don't want
```
removed

mhevery commented on packages/core/src/util/array_utils.ts:
```diff
@@ -24,22 +24,62 @@ export function addAllToArray(items: any[], arr: any[]) {
+ * @param dst Destination array which can have existing values.
+ * @returns `true` if `dst` array was muted in order to represent flatten version of `src`.
+ */
+export function flattenIntoExisting(src: any[], dst: any[]): boolean {
```
fixed

mhevery commented on packages/core/src/util/array_utils.ts:
```diff
@@ -24,22 +24,62 @@ export function addAllToArray(items: any[], arr: any[]) {
+ * @returns Location where the diffing ended in `dst` array. If `<0` then `~value` is `dstIdx` and a
+ *     `dst` array was mutated (change detected).
+ */
+function flattenIntoExistingRecursive(src: any[], dst: any[], dstIdx: number): number {
```
fixed

mhevery commented on packages/core/test/util/array_utils_spec.ts:
```diff
@@ -28,6 +28,44 @@ describe('array_utils', () => {
   });
 
+  describe('flattenIntoExisting', () => {
+    it('should detectNoChanges on empty set', () => {
```
fixed

mhevery commented on packages/core/test/util/array_utils_spec.ts:
```diff
@@ -28,6 +28,44 @@ describe('array_utils', () => {
+      expect(dst).toEqual(['a', 'b']);
+    });
+
+    it('should detectNoChanges on nested value', () => {
```
fixed

mhevery commented:
@IgorMinar  `Array#flat` seems like a lot of work, so I am going to do it in a separate PR since it includes changes to polyfills and I don't think we should be changing polyfills in patch release.

AndrewKushnir commented on packages/core/src/linker/query_list.ts:
```diff
@@ -135,28 +169,36 @@ export class QueryList<T> implements Iterable<T> {
-  /** internal */
+  /** @internal */
   destroy(): void {
     (this.changes as EventEmitter<any>).complete();
     (this.changes as EventEmitter<any>).unsubscribe();
```
We should probably do that for both `_changesDeprecated` and `_changesStrict` event emitters (in case both emitters were used in an app).

AndrewKushnir commented on packages/core/src/linker/query_list.ts:
```diff
@@ -135,28 +169,36 @@ export class QueryList<T> implements Iterable<T> {
-    (this as {last: T}).last = this._results[this.length - 1];
-    (this as {first: T}).first = this._results[0];
+    if (flattenIntoExisting(resultsTree, this._results)) {
+      (this as {length: number}).length = this._results.length;
+      (this as {last: T}).last = this._results[this.length - 1];
+      (this as {first: T}).first = this._results[0];
```
This is outside of the context of this PR, but it feels like these fields should be just getters on the `QueryList` class (so setting their values would not be required) and that should also help [with typings above](https://github.com/angular/angular/pull/40091/files#diff-28d6add596d41a073a75f60a25bddf1655300ea5afc2b09b8bced28f64f02675R52).

AndrewKushnir commented:
FYI, started [global presubmit](https://test.corp.google.com/ui#id=OCL:347540875:BASE:347540881:1608013027172:a7338ace).

JoostK commented on packages/core/src/linker/query_list.ts:
```diff
@@ -135,28 +169,36 @@ export class QueryList<T> implements Iterable<T> {
    *
    * @param resultsTree The query results to store
+   * @returns `true` only if the new `resultTree` resulted in a different query set.
+   * @internal
```
Although this is marked `@internal`, it still shows up in the public api. As this is targeting patch I don't think we can/should hide this method in a patch release.

mhevery replied:
This is an internal method which should never be called by anyone. I can take it out if you feel strongly about it.

JoostK replied:
So I'm fine with marking it internal, it's just that doing so could break projects who call this method manually. So we may need to go through a deprecation period.

JoostK commented on packages/core/src/linker/query_list.ts:
```diff
@@ -45,13 +45,47 @@ function symbolIterator<T>(this: QueryList<T>): Iterator<T> {
+   * NOTE: This is a proper implementation which only fires if there are underlying changes in the
+   * `Observable` (the previous implementation would fire even if no changes were detected.)
+   */
+  get changesStrict(): Observable<QueryList<T>> {
```
Is the intention to mark this deprecated when `changes` is switched from `changesDeprecated` to `changesString`?

mhevery replied:
yes

JoostK reviewed (commented):
I'd personally mark this for minor and not patch, since it introduces new API.

mhevery commented:
[presubmit](http://test/OCL:347030270:BASE:347707813:1608075767113:10bcfa78)

AndrewKushnir commented:
[Global presubmit #2](https://test.corp.google.com/OCL:347751946:BASE:347752135:1608096829955:af3384b2) (using Ivy).

mhevery commented:
[presubmit](http://test/OCL:347030270:BASE:347912969:1608162502638:af74c34a)

jelbourn commented on packages/core/src/linker/query_list.ts:
```diff
@@ -45,13 +45,48 @@ function symbolIterator<T>(this: QueryList<T>): Iterator<T> {
+   *
+   * NOTE: This currently points to `changesDeprecated` which incorrectly notifies of changes even
+   * if no changes to `QueryList` have occurred. (It fires more often than it needs to.)
+   * The implementation will change to point `changesStrict` starting with v12 (or later.)
```
I'm not a fan of the "or later" here. Do we know how breaking changing the default will be in Google? That will tell us if it can fit in the scope for v12 or if we should slate it for v13.

mhevery replied:
This issue is fixing difference in behavior in VE vs Ivy. Rather than getting Ivy to behave in same broken way as VE we are trying to fix this. Turns out that there are existing projects which rely on the broken behavior and there are some failures. 

Do you have suggestions on how to change the wording?

jelbourn replied:
If we can feasibly change the default in Google by v12, this should just say "v12". If not, it should be "v13", either way omitting the "(or later)" since it adds uncertainty.

mhevery replied:
fixed

jelbourn commented on packages/core/src/util/array_utils.ts:
```diff
@@ -20,6 +20,31 @@ export function addAllToArray(items: any[], arr: any[]) {
+ *
+ * @param a first array
+ * @param b second array
+ * @param predicate Optional functions for extracting stable object identity from a value in the
```
This doesn't match the definition of "predicate", which is a function that returns a boolean. Following Angular's convention, I think this would be called `trackBy`, but since this is internal I might use `identityAccessor`.

mhevery replied:
changed.

jelbourn commented on packages/core/src/util/array_utils.ts:
```diff
@@ -20,6 +20,31 @@ export function addAllToArray(items: any[], arr: any[]) {
+export function arrayEquals<T>(a: T[], b: T[], predicate?: (value: T) => unknown): boolean {
+  if (a.length !== b.length) return false;
+  for (let i = 0; i < a.length; i++) {
+    let srcValue = b[i];
```
I think using `src` and `dst` is a bit odd here, since it's projecting meaning in a generic utility function. 'd lean towards just `valueA` / `valueB`.

mhevery replied:
fixed

jelbourn commented:
Also, separately from code review- is there something that makes this a priority? What's the impact of the extra stream emits? Since cleaning this up will require a large scale change inside Google, we should be deliberate about investing our time in this.

mhevery commented:
> Also, separately from code review- is there something that makes this a priority? What's the impact of the extra stream emits? Since cleaning this up will require a large scale change inside Google, we should be deliberate about investing our time in this.

Yes, we need to land this to unblock code in `g3`

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

IgorMinar reviewed (commented):
as discussed on slack, let's go with the `@ContentChildren('foo', {strictChangeEmit: true}) foos!: QueryList<...>;`

mhevery committed:
refactor(core): Remove the need for explicit static query instruction

Because the query now has `flags` which specify the mode, the static query
instruction can now be remove. It is simply normal query with `static` flag.

AndrewKushnir commented on packages/core/src/linker/query_list.ts:
```diff
@@ -135,20 +151,27 @@ export class QueryList<T> implements Iterable<T> {
-    (this as {last: T}).last = this._results[this.length - 1];
-    (this as {first: T}).first = this._results[0];
+  reset(resultsTree: Array<T|any[]>, identityAccessor?: (value: T) => unknown): void {
+    const self = this as QueryListInternal<T>;
```
May be add a quick comment on why we use `self` vs directly using `this`.

mhevery replied:
denu

AndrewKushnir commented on packages/core/src/linker/query_list.ts:
```diff
@@ -135,20 +151,27 @@ export class QueryList<T> implements Iterable<T> {
-    (this as {first: T}).first = this._results[0];
+  reset(resultsTree: Array<T|any[]>, identityAccessor?: (value: T) => unknown): void {
+    const self = this as QueryListInternal<T>;
+    (self as {dirty: boolean}).dirty = false;
```
I guess the `dirty` field can be added to the `QueryListInternal` and this cast would not be needed.

mhevery replied:
I tried it and it does not work. :(

AndrewKushnir commented on packages/core/src/render3/interfaces/query.ts:
```diff
@@ -18,9 +18,39 @@ import {TView} from './view';
+export const enum QueryFlags {
+  /**
+   * No flags
+   */
+
+  none = 0,
+  /**
```
```suggestion
   */
  none = 0,
  
  /**
```

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/render3/query.ts:
```diff
@@ -489,50 +473,27 @@ function viewQueryInternal<T>(
+    createTQuery(tView, new TQueryMetadata_(predicate, flags, read), tNode.index);
     saveContentQueryAndDirectiveIndex(tView, directiveIndex);
-    if (isStatic) {
+    if (flags & QueryFlags.isStatic) {
```
Should this be similar to other places?
```suggestion
    if ((flags & QueryFlags.isStatic) === QueryFlags.isStatic) {
```

mhevery replied:
done

AndrewKushnir reviewed (commented):
Thanks for the fix @mhevery 👍 Just left a few comments.

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/queries/static_content_query.js:
```diff
@@ -3,8 +3,8 @@ ContentQueryComponent.ɵcmp = $r3$.ɵɵdefineComponent({
     if (rf & 1) {
-    $r3$.ɵɵstaticContentQuery(dirIndex, SomeDirective, true);
-    $r3$.ɵɵcontentQuery(dirIndex, $ref0$, true);
+      $r3$.ɵɵcontentQuery(dirIndex, SomeDirective, 3);
```
The `3` here can be expressed as a union if we declare a union flag matcher for it here:

https://github.com/angular/angular/blob/76ae87406f487adc2f3a501ca10672a7d6cde49e/packages/compiler-cli/test/compliance/test_helpers/expected_file_macros.ts#L36-L37

Doing so would allow this to read

```suggestion
      $r3$.ɵɵcontentQuery(dirIndex, SomeDirective, __QueryFlags.descendants__|__QueryFlags.isStatic__);
```

This applies to all compliance tests under `packages/compiler-cli/test/compliance` (but not compliance_old nor ngtsc/ngcc tests)

mhevery replied:
fixed

JoostK commented on packages/core/src/linker/query_list.ts:
```diff
@@ -135,20 +151,27 @@ export class QueryList<T> implements Iterable<T> {
    */
   notifyOnChanges(): void {
-    (this.changes as EventEmitter<any>).emit(this);
+    if (this._changes && (this._emitDistinctChangesOnly ? this._changesDetected : true))
```
This should be logically equivalent:

```suggestion
    if (this._changes && (this._changesDetected || !this._emitDistinctChangesOnly))
```

mhevery replied:
fixed

JoostK commented on packages/core/src/linker/query_list.ts:
```diff
@@ -169,3 +192,14 @@ export class QueryList<T> implements Iterable<T> {
+/**
+ * Internal set of APIs used by the framework. (not to be made public)
+ */
+export interface QueryListInternal<T> extends QueryList<T> {
```
```suggestion
interface QueryListInternal<T> extends QueryList<T> {
```

mhevery replied:
fixed

JoostK commented on packages/core/src/render3/query.ts:
```diff
@@ -427,14 +428,15 @@ export function ɵɵqueryRefresh(queryList: QueryList<any>): boolean {
 
   const tQuery = getTQuery(tView, queryIndex);
-  if (queryList.dirty && (isCreationMode(lView) === tQuery.metadata.isStatic)) {
+  const isStatic = (tQuery.metadata.flags & QueryFlags.isStatic) === QueryFlags.isStatic;
```
This used to be inline, allowing the computation to be skipped if `!queryList.dirty`. Since this executes in each CD cycle, maybe it matters for perf (although the same already applies to the preexisting `getTQuery` call)

mhevery replied:
inlined

JoostK commented on packages/core/src/render3/query.ts:
```diff
@@ -202,7 +202,8 @@ class TQuery_ implements TQuery {
 
   private isApplyingToNode(tNode: TNode): boolean {
-    if (this._appliesToNextNode && this.metadata.descendants === false) {
+    const isDescend = (this.metadata.flags & QueryFlags.descendants) === QueryFlags.descendants;
+    if (this._appliesToNextNode && !isDescend) {
```
Same comment here about the observation that `isDescend` is now computed even when `this._appliesToNextNode` is false. But I believe this only executes on first template pass so it's less perf sensitive.

mhevery replied:
inlined

JoostK commented on packages/core/src/util/array_utils.ts:
```diff
@@ -20,6 +20,31 @@ export function addAllToArray(items: any[], arr: any[]) {
+ *
+ * @param a first array
+ * @param b second array
+ * @param identityAccessor Optional functions for extracting stable object identity from a value in
```
```suggestion
 * @param identityAccessor Optional function for extracting stable object identity from a value in
```

mhevery replied:
done

JoostK commented on packages/core/test/acceptance/query_spec.ts:
```diff
@@ -1123,6 +1123,47 @@ describe('query logic', () => {
+    onlyInIvy('ViewEngine behavior does not change')
+        .it('should only fire if the content of the query changes', () => {
+          // When views are inserted/removed the content query need to be recomputed.
+          // Recomputing query may result in no changes to the query (the item added/remove was not
```
```suggestion
          // Recomputing the query may result in no changes to the query (the item added/removed was not
```

mhevery replied:
done

JoostK commented on packages/core/test/render3/jit/declare_component_spec.ts:
```diff
@@ -128,16 +128,17 @@ describe('component declaration jit compilation', () => {
 
-        // "byToken" should use `staticContentQuery` with `true` for descendants flag and
-        // `ElementRef` as read token, and bind to the first result in the query result.
+        // "byToken" should use `staticContentQuery` with `3`
+        // (`QueryFlags.descendants|QueryFlags.isStatic`) for descendants flag and `ElementRef` as
```
```suggestion
        // "byToken" should use `contentQuery` with `3`
        // (`QueryFlags.descendants|QueryFlags.isStatic`) for query flag and `ElementRef` as
```

mhevery replied:
fixed

JoostK commented on packages/core/test/render3/jit/declare_component_spec.ts:
```diff
@@ -168,13 +169,14 @@ describe('component declaration jit compilation', () => {
         // "byRef" should use `viewQuery` with `false` for descendants flag without a read token,
```
```suggestion
        // "byRef" should use `viewQuery` with `0` (`QueryFlags.none`) for query flag without a read token,
```

mhevery replied:
fixed

JoostK commented on packages/core/test/render3/jit/declare_component_spec.ts:
```diff
@@ -128,16 +128,17 @@ describe('component declaration jit compilation', () => {
       contentQueries: functionContaining([
-        // "byRef" should use `contentQuery` with `false` for descendants flag without a read token,
-        // and bind to the full query result.
+        // "byRef" should use `contentQuery` with `0` (`QueryFlags.none`) for descendants flag
```
```suggestion
        // "byRef" should use `contentQuery` with `0` (`QueryFlags.none`) for query flag
```

mhevery replied:
fixed

JoostK commented on packages/core/test/render3/jit/declare_component_spec.ts:
```diff
@@ -168,13 +169,14 @@ describe('component declaration jit compilation', () => {
-        // "byToken" should use `staticViewQuery` with `true` for descendants flag and
-        // `ElementRef` as read token, and bind to the first result in the query result.
+        // "byToken" should use `viewQuery` with `3`
+        // (`QueryFlags.descendants|QueryFlags.isStatic`) for descendants flag and `ElementRef` as
```
```suggestion
        // (`QueryFlags.descendants|QueryFlags.isStatic`) for query flag and `ElementRef` as
```

mhevery replied:
fixed

JoostK commented on packages/core/test/render3/jit/declare_directive_spec.ts:
```diff
@@ -101,16 +101,17 @@ describe('directive declaration jit compilation', () => {
       contentQueries: functionContaining([
-        // "byRef" should use `contentQuery` with `false` for descendants flag without a read token,
-        // and bind to the full query result.
+        // "byRef" should use `contentQuery` with `0` (`QueryFlags.descendants|QueryFlags.isStatic`)
+        // for descendants flag without a read token, and bind to the full query result.
```
```suggestion
        // "byRef" should use `contentQuery` with `0` (`QueryFlags.none`)
        // for query flag without a read token, and bind to the full query result.
```

mhevery replied:
fixed

JoostK commented on packages/core/test/render3/jit/declare_directive_spec.ts:
```diff
@@ -101,16 +101,17 @@ describe('directive declaration jit compilation', () => {
-        // "byToken" should use `staticContentQuery` with `true` for descendants flag and
-        // `ElementRef` as read token, and bind to the first result in the query result.
+        // "byToken" should use `viewQuery` with `3` (`QueryFlags.static|QueryFlags.descendants`)
+        // for descendants flag and `ElementRef` as read token, and bind to the first result in the
```
```suggestion
        // for query flag and `ElementRef` as read token, and bind to the first result in the
```

mhevery replied:
fixed

JoostK commented on packages/core/test/render3/jit/declare_directive_spec.ts:
```diff
@@ -140,13 +141,14 @@ describe('directive declaration jit compilation', () => {
         // "byRef" should use `viewQuery` with `false` for descendants flag without a read token,
```
```suggestion
        // "byRef" should use `viewQuery` with `0` (`QueryFlags.none`) for query flag without a read token,
```

mhevery replied:
fixed

JoostK commented on packages/core/test/render3/jit/declare_directive_spec.ts:
```diff
@@ -140,13 +141,14 @@ describe('directive declaration jit compilation', () => {
-        // "byToken" should use `staticViewQuery` with `true` for descendants flag and
-        // `ElementRef` as read token, and bind to the first result in the query result.
+        // "byToken" should use `viewQuery` with `3` (`QueryFlags.static|QueryFlags.descendants`)
+        // for descendants flag and `ElementRef` as read token, and bind to the first result in the
```
```suggestion
        // for query flag and `ElementRef` as read token, and bind to the first result in the
```

mhevery replied:
fixed

mhevery commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -435,13 +436,24 @@ function selectorsFromGlobalMetadata(
+ */
+function toQueryFlags(query: R3QueryMetadata): number {
+  // NOTE: Verify that changes here match
+  return (query.descendants ? 1 /* TQueryFlags.descendants */ : 0) |
```
fixed

mhevery commented on packages/core/src/linker/element_ref.ts:
```diff
@@ -86,3 +86,14 @@ export class ElementRef<T = any> {
+/**
+ * Unwraps `ElementRef` and return the `nativeElement`.
+ *
+ * Conditionally unwrap the `ElementRef`.
```
fixed

mhevery commented on packages/core/src/linker/query_list.ts:
```diff
@@ -45,15 +45,31 @@ function symbolIterator<T>(this: QueryList<T>): Iterator<T> {
-  readonly first!: T;
-  // TODO(issue/24571): remove '!'.
-  readonly last!: T;
+  readonly first: T = undefined!;
+  readonly last: T = undefined!;
```
If they are uninitialized than the object changes shape later, which can cause megamorphic reads. The change does not change runtime behavior, the only thing it does is to explicitly initializes the fields so that we don't have class transition later.

mhevery commented on packages/core/src/linker/query_list.ts:
```diff
@@ -45,15 +45,31 @@ function symbolIterator<T>(this: QueryList<T>): Iterator<T> {
+  /**
+   * Returns `Observable` of `QueryList` notifying the subscriber of changes.
+   *
+   * NOTE: This currently points to `changesDeprecated` which incorrectly notifies of changes even
+   * if no changes to `QueryList` have occurred. (It fires more often than it needs to.)
+   * The implementation will change to point `changesStrict` starting with v12.
```
deleted

mhevery commented on packages/core/src/linker/query_list.ts:
```diff
@@ -45,15 +45,31 @@ function symbolIterator<T>(this: QueryList<T>): Iterator<T> {
+   * if no changes to `QueryList` have occurred. (It fires more often than it needs to.)
+   * The implementation will change to point `changesStrict` starting with v12.
+   */
+  get changes(): Observable<any> {
+    return this._changes || (this._changes = new EventEmitter());
+  }
```
I think this is better. Most of the time do not subscribe to QueryList, but we still allocate an object which no one uses. Doing it lazily is memory pressure improvement.

AndrewKushnir replied:
Agree, that makes sense, thanks!

mary-poppins commented:
You can preview 1d4e830 at https://pr40091-1d4e830.ngbuilds.io/.
You can preview f02babb at https://pr40091-f02babb.ngbuilds.io/.
You can preview 2cd900f at https://pr40091-2cd900f.ngbuilds.io/.
You can preview 266fa09 at https://pr40091-266fa09.ngbuilds.io/.
You can preview 8a377c1 at https://pr40091-8a377c1.ngbuilds.io/.
You can preview 278dbce at https://pr40091-278dbce.ngbuilds.io/.
You can preview d7c49d4 at https://pr40091-d7c49d4.ngbuilds.io/.
You can preview ce929a5 at https://pr40091-ce929a5.ngbuilds.io/.
You can preview bdb1b09 at https://pr40091-bdb1b09.ngbuilds.io/.
You can preview 625551f at https://pr40091-625551f.ngbuilds.io/.
You can preview e4dcceb at https://pr40091-e4dcceb.ngbuilds.io/.
You can preview f777d24 at https://pr40091-f777d24.ngbuilds.io/.
You can preview 1e8a1d8 at https://pr40091-1e8a1d8.ngbuilds.io/.
You can preview 6ff6cac at https://pr40091-6ff6cac.ngbuilds.io/.
You can preview acf73d7 at https://pr40091-acf73d7.ngbuilds.io/.

mhevery commented on packages/core/src/linker/query_list.ts:
```diff
@@ -135,20 +151,27 @@ export class QueryList<T> implements Iterable<T> {
    * occurs.
    *
    * @param resultsTree The query results to store
+   * @param identityAccessor Optional functions for extracting stable object identity from a value
+   *     in the array.
```
fixed

mhevery commented on packages/core/src/linker/query_list.ts:
```diff
@@ -45,15 +45,31 @@ function symbolIterator<T>(this: QueryList<T>): Iterator<T> {
+  }
+
+  /**
+   * @param emitDistinctChangesOnly Whether `QueryList.changes` should fire only when actual change
```
it does.

IgorMinar reviewed (changes requested):
I like the new api name `emitDistinctChangesOnly` a lot more than the previous options. Can you please rerun the api guard to get the PR into a green state? thanks!

mary-poppins commented:
You can preview 77eb52b at https://pr40091-77eb52b.ngbuilds.io/.

mhevery commented on goldens/public-api/core/core.d.ts:
```diff
@@ -1,261 +1,243 @@
 export declare interface AbstractType<T> extends Function {
```
fixed

mary-poppins commented:
You can preview 847658a at https://pr40091-847658a.ngbuilds.io/.

mhevery commented:
[presubmit](http://test/OCL:347030270:BASE:351188007:1610390777230:e83a0872)

AndrewKushnir commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -435,13 +436,54 @@ function selectorsFromGlobalMetadata(
+  /**
+   * No flags
+   */
+
+  none = 0,
```
Small indentation update (rm line before `none`, add one after):
```suggestion
  none = 0,

```

mhevery replied:
Why line after? I think it is easier to read. Other places have line breaks like that as well.

AndrewKushnir replied:
It just looks like the comment is not "attached" to the `none = 0` string (since there is a line in between a comment and an expression).

mhevery replied:
yes, that one (break between comment and declaration) I removed. But you are suggesting to remove the empty line after. That one I think should stay.

AndrewKushnir replied:
Yes, I agree that we should keep an empty line after 👍

mary-poppins commented:
You can preview c469e9a at https://pr40091-c469e9a.ngbuilds.io/.

mary-poppins commented:
You can preview 9c15a80 at https://pr40091-9c15a80.ngbuilds.io/.

mary-poppins commented:
You can preview f2f3cfb at https://pr40091-f2f3cfb.ngbuilds.io/.

JoostK commented:
Is this still intended to target patch, given that it introduces new API? Since 11.1 RC is imminent, wouldn't it be better to introduce it there?

mhevery commented:
> Is this still intended to target patch, given that it introduces new API? Since 11.1 RC is imminent, wouldn't it be better to introduce it there?

What advantage will that bring?

JoostK commented:
> > Is this still intended to target patch, given that it introduces new API? Since 11.1 RC is imminent, wouldn't it be better to introduce it there?
> 
> What advantage will that bring?

Avoid the introduction of public API in a patch release?

mary-poppins commented:
You can preview e3949bc at https://pr40091-e3949bc.ngbuilds.io/.

mhevery commented:
> Avoid the introduction of public API in a patch release? Or would you consider this acceptable as it's solely an additive change?

My thinking is that this is a bug fix, which just happens to be behind a flag. But I can be convinced to do something else.

mary-poppins commented:
You can preview e26d05a at https://pr40091-e26d05a.ngbuilds.io/.

alxhub commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -232,6 +232,14 @@ export interface R3DeclareQueryMetadata {
 
+  /**
+   * True to only fire changes if there are underlying changes to the query. This will become `true`
+   * be default in v12. `QueryList.changes` would fire even if no changes to the query list were
```
Perhaps convert "This will become true by default in v12" to a TODO so we don't have version number references in doc strings directly.

mhevery replied:
done

alxhub commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -232,6 +232,14 @@ export interface R3DeclareQueryMetadata {
+   * detected. This is not ideal, as changes should only fire if the `QueryList` actually materially
+   * changed.
+   */
+  emitDistinctChangesOnly?: boolean;
```
```suggestion
  emitDistinctChangesOnly: boolean;
```

In the compiler we try to avoid optional fields, especially optional `boolean`s. (I know `descendants` above is in violation of this rule - it should really be non-optional too). Prefer explicitly passing the property.

JoostK replied:
This one is special ;) This interface documents the structure as generated by the linker, where we omit properties when they are `false` (or another typical default).

alxhub replied:
That brings up an interesting question though. If the plan is to change this property from default-false to default-true in a future version, how will that affect prelinked code already on NPM (hypothetically assuming such code exists)? Should we preemptively always emit the property and set it explicitly to `false`, with the expectation that a missing value will eventually be interpreted as `true`? Or do we have some other way of changing the default without affecting existing prelinked code?

JoostK replied:
We discussed this a while ago but I don't recall the exact outcome. But yes, the linker's defaults are never meant to be switched, as that would affect published code.

Yesterday we agreed to start a document with public API considerations, I'll incorporate this aspect as well so we can reassess (the plan is to present that document at next week's sync)

alxhub replied:
Ty! Looking forward to it. This code LGTM for now

alxhub commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -435,13 +436,54 @@ function selectorsFromGlobalMetadata(
+/**
+ * A set of flags to be used with Queries.
+ *
+ * NOTE: Ensure changes here are reflected in `packages/compiler/src/render3/view/compiler.ts`
```
This comment refers to the current file - perhaps a copy/paste error?

mhevery replied:
fixed

alxhub commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -435,13 +436,54 @@ function selectorsFromGlobalMetadata(
+  /**
+   * Whether or not the query should descend into children.
+   */
+  descendants = 1,
```
I tend to like using `0x0001`, `0x0002`, etc. notation for bitfield enum constants, to make it more obvious that these are intended to be specific masked bits.

mhevery replied:
Good point, done.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api
Reviewed-for: size-tracking

JoostK reviewed (approved):
LGTM; I'll follow up with some additional tests for the linker, but that's not blocking here.

mary-poppins commented:
You can preview f08aa33 at https://pr40091-f08aa33.ngbuilds.io/.

IgorMinar commented on packages/core/src/metadata/di.ts:
```diff
@@ -268,6 +281,10 @@ export interface ViewChildrenDecorator {
    *
    * * **selector** - The directive type or the name used for querying.
    * * **read** - Used to read a different token from the queried elements.
+   * * **emitDistinctChangesOnly** - True to only fire changes if there are underlying changes to
+   *   the query. This will become `true` be default in v12. `QueryList.changes` would fire even if
+   * no changes to the query list were detected. This is not ideal, as changes should only fire if
+   *   the `QueryList` actually materially changed.
```
Consider something simpler like (might require reformatting):

```suggestion
   * * **emitDistinctChangesOnly** - The ` QueryList#changes` observable will emit new values only if the QueryList result has changed. 
   *     The default value will change from `false` to `true` in v12.
   *     When `false` the `changes` observable might emit even if the QueryList has not changed.
```

IgorMinar reviewed (approved):
LGTM with some suggested changes to docs.

Reviewed-for: public-api

IgorMinar commented:
please reformat the code as well to fix the linting issue.

mhevery commented:
[presubmit](http://test/OCL:347030270:BASE:351472779:1610497660458:e9897b85)

mary-poppins commented:
You can preview af6d6cd at https://pr40091-af6d6cd.ngbuilds.io/.

IgorMinar reviewed (approved):
LGTM 

Reviewed-for: size-tracking, public-api

mary-poppins commented:
You can preview 29c4f20 at https://pr40091-29c4f20.ngbuilds.io/.

mary-poppins commented:
You can preview 469090e at https://pr40091-469090e.ngbuilds.io/.

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts:
```diff
@@ -147,6 +147,8 @@ function toQueryMetadata<TExpression>(obj: AstObject<R3DeclareQueryMetadata, TEx
     predicate,
     descendants: obj.has('descendants') ? obj.getBoolean('descendants') : false,
+    emitDistinctChangesOnly:
+        obj.has('emitDistinctChangesOnly') ? obj.getBoolean('emitDistinctChangesOnly') : false,
```
The `false` needs to be flipped here. And we may want to include a compliance test which excercies both values so we can be certain that the linker roundtrip is successful

mhevery replied:
done

mary-poppins commented:
You can preview 4d81a44 at https://pr40091-4d81a44.ngbuilds.io/.

mary-poppins commented:
You can preview 29cd0de at https://pr40091-29cd0de.ngbuilds.io/.

AndrewKushnir commented:
@mhevery this PR was merged into RC (11.1.x) and master branches, but there was a conflict of merging it to the patch branch (11.0.x). Could you please create a patch-only version of this change if needed? Thank you.

JoostK commented:
> @mhevery this PR was merged into RC (11.1.x) and master branches, but there was a conflict of merging it to the patch branch (11.0.x). Could you please create a patch-only version of this change if needed? Thank you.

I believe RC should be sufficient; there's no real need for this to land in 11.0.x with 11.1 just around the corner (there might not even be another 11.0.x release?) and porting this is non-trivial as the linker code doesn't exist in 11.0.x.

AndrewKushnir commented:
Thanks @JoostK. Right, we also discussed that with @mhevery and agreed to keep this change in RC and master only (no changes to the patch branch).
